hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OmRequestGatekeeper.java

```diff
@@ -0,0 +1,144 @@
+ * implementation uses 3 bins of striped locks, We can change the lock management implementation later as needed with
+ * little to no effect on any other classes.
+ */
+public class OmRequestGatekeeper {
+  private final Striped<ReadWriteLock> volumeLocks;
+  private final Striped<ReadWriteLock> bucketLocks;
+  // Currently only key level write locks are required. This lets us use the bulkGet API from the Striped class.
+  private final Striped<Lock> keyLocks;
+  private final long timeoutMillis;
+
+  // TODO Arbitrary values for now.
+  private static final int NUM_VOLUME_STRIPES = 128;
+  private static final int NUM_BUCKET_STRIPES = 1024;
+  private static final int NUM_KEY_STRIPES = 4096;
+
+  // If stats about the locks need to be added, they should be tracked here in the gatekeeper.
+
+  public OmRequestGatekeeper(long timeoutMillis) {
+    this.timeoutMillis = timeoutMillis;
+    volumeLocks = Striped.readWriteLock(NUM_VOLUME_STRIPES);
+    bucketLocks = Striped.readWriteLock(NUM_BUCKET_STRIPES);
+    keyLocks = Striped.lock(NUM_KEY_STRIPES);
+  }
+
+  /**
+   * Acquires all the locks specified by the {@code lockInfo} parameter. Caller is responsible for releasing the locks
+   * by closing the returned object. If an exception is thrown, no locks will be acquired when the method exits.
+   * @param lockInfo Defines the locks to acquire.
+   * @return A handle that can be closed to release the acquired locks.
+   * @throws InterruptedException If the thread is interrupted while waiting to acquire some of the locks.
+   * @throws TimeoutException If the timeout elapses while waiting to acquire any of the locks individually.
+   */
+  public AutoCloseable lock(OmLockInfo lockInfo) throws InterruptedException, TimeoutException {
+    // Common case is 1 volume, 1 bucket, and at most 2 key locks.
+    List<Lock> locks = new ArrayList<>(4);
+
+    Optional<LockInfo> optionalVolumeLock = lockInfo.getVolumeLock();
+    Optional<LockInfo> optionalBucketLock = lockInfo.getBucketLock();
+    Optional<Set<LockInfo>> optionalKeyLocks = lockInfo.getKeyLocks();
+
+    if (optionalVolumeLock.isPresent()) {
+      LockInfo volumeLockInfo = optionalVolumeLock.get();
+      if (volumeLockInfo.isWriteLock()) {
+        locks.add(volumeLocks.get(volumeLockInfo.getKey()).writeLock());
+      } else {
+        locks.add(volumeLocks.get(volumeLockInfo.getKey()).readLock());
+      }
+    }
+
+    if (optionalBucketLock.isPresent()) {
+      LockInfo bucketLockInfo = optionalBucketLock.get();
+      if (bucketLockInfo.isWriteLock()) {
+        locks.add(bucketLocks.get(bucketLockInfo.getKey()).writeLock());
+      } else {
+        locks.add(bucketLocks.get(bucketLockInfo.getKey()).readLock());
+      }
+    }
+
+    if (optionalKeyLocks.isPresent()) {
+      for (Lock keyLock: keyLocks.bulkGet(optionalKeyLocks.get())) {
+        locks.add(keyLock);
+      }
+    }
+
+    acquireLocks(locks);
+    return () -> releaseLocks(locks);
+  }
```

Comment:
The [lock overhead](https://en.wikipedia.org/wiki/Lock_(computer_science)#Granularity) probably is quite high in this method.